zio-http/shared/src/main/scala/zio/http/internal/StringSchemaCodec.scala

```diff
@@ -0,0 +1,658 @@
+package zio.http.internal
+
+import java.time._
+import java.util.{Currency, UUID}
+
+import scala.annotation.tailrec
+import scala.util.Try
+
+import zio.{Cause, Chunk, Unsafe}
+
+import zio.schema.codec.DecodeError
+import zio.schema.validation.{Validation, ValidationError}
+import zio.schema.{Schema, StandardType, TypeId}
+
+import zio.http.codec.HttpCodecError
+import zio.http.internal.StringSchemaCodec.{PrimitiveCodec, decodeAndUnwrap, emptyStringIsValue, validateDecoded}
+
+private[http] final case class ErrorConstructor(
+  missing: String => HttpCodecError,
+  missingAll: Chunk[String] => HttpCodecError,
+  invalid: Chunk[ValidationError] => HttpCodecError,
+  malformed: (String, DecodeError) => HttpCodecError,
+  invalidCount: (String, Int, Int) => HttpCodecError,
+)
+
+private[http] final case class StringSchemaCodec[A, Target](
```

Comment:
case classes that store functions are generally better off as `trait`s (interfaces implemented with anonymous classes).